components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs

```diff
@@ -440,32 +458,72 @@ public class LegacyJobOrchestrator : IJobOrchestrator, IPublisherConfigServices,
             _agentConfig.TriggerConfigUpdate(this, new EventArgs());
         }
 
+        private void ValidateRequest(PublishedNodesEntryModel request) {
+            var requestList = new List<PublishedNodesEntryModel> { request };
+            var requestJson = _jsonSerializer.SerializeToString(requestList);
+
+            // This will throw SerializerException if values of request fields are not conformant.
+            DeserializePublishedNodes(requestJson);
+        }
+
+        /// <summary>
+        /// Persist _publishedNodesEntries to published nodes file.
+        /// </summary>
+        private void PersistPublishedNodes() {
+            var updatedContent = _jsonSerializer.SerializeToString(_publishedNodesEntries, SerializeOption.Indented);
```

Comment:
>_publishedNodesEntries [](http://example.com/codeflow?start=67&length=22)

I think we should put this in a _lockConfig block

Reply:
All calls to `PersistPublishedNodes()` are happening from direct method handlers which acquire `_lockConfig` lock. And it is a private method, so no calls from outside of the class are permitted.

Reply:
well, I still believe the code inside a method (private or not) must not depend on external lock blocks. The code might be changed, refactored and such a design could easily lead to deadlocks or sync issues that are super difficult to debug and fix

Reply:
I will disagree here as the lock are needed for logical blocks where we change or access `_publishedNodesEntries`. This method is a part of that block just separated in helper method. I think that we should be able to have internal methods that assume being called under a lock. Of course that expectation should be documented  and enforced during reviews.

The other problem with the proposal is that we cannot achieve that with current locks as they are not reentrant, meaning that even from within the same thread the lock cannot be acquired twice. And that would be necessary as we have already acquired the lock from the calling block. Releasing the lock in between is not an option here as that introduces possibilities of inconsistent changes to `_publishedNodesEntries` and its serialized version in JSON file. We can use reentrant locks such as [Monitor](https://docs.microsoft.com/en-us/dotnet/api/system.threading.monitor?redirectedfrom=MSDN&view=net-6.0) here to work around that, but I would still argue that having internal methods that assume being called from within a lock block is a valid option.

Reply:
Got it. It's fine with me.